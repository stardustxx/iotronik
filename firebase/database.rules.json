{
  "rules": {
    ".read": "auth != null",
    ".write": "auth != null",
    "incident": {
      ".indexOn": ["time"]
    }
  }
}
