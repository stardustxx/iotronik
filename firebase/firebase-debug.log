[debug] ----------------------------------------------------------------------
[debug] Command:      /usr/local/Cellar/node/7.10.0/bin/node /usr/local/bin/firebase serve
[debug] CLI Version:  3.7.0
[debug] Platform:     darwin
[debug] Node Version: v7.10.0
[debug] Time:         Wed May 10 2017 11:19:10 GMT-0700 (PDT)
[debug] ----------------------------------------------------------------------
[debug] 
[debug] > command requires scopes: ["email","openid","https://www.googleapis.com/auth/cloudplatformprojects.readonly","https://www.googleapis.com/auth/firebase","https://www.googleapis.com/auth/cloud-platform"]
[debug] > refreshing access token with scopes: ["email","https://www.googleapis.com/auth/cloud-platform","https://www.googleapis.com/auth/cloudplatformprojects.readonly","https://www.googleapis.com/auth/firebase","openid"]
[debug] >>> HTTP REQUEST POST https://www.googleapis.com/oauth2/v3/token { refresh_token: '1/ytA7LoLSmKea1CC27yRzeq9ryKjoeZsiUI3nM3RcvN8',
  client_id: '563584335869-fgrhgmd47bqnekij5i8b5pr03ho849e6.apps.googleusercontent.com',
  client_secret: 'j9iVZfS8kkCEFUPaAeJV0sAi',
  grant_type: 'refresh_token',
  scope: 'email https://www.googleapis.com/auth/cloud-platform https://www.googleapis.com/auth/cloudplatformprojects.readonly https://www.googleapis.com/auth/firebase openid' } 
 Wed May 10 2017 11:19:10 GMT-0700 (PDT)
[debug] <<< HTTP RESPONSE 200
[debug] >>> HTTP REQUEST GET https://admin.firebase.com/v1/projects/iotroniks  
 Wed May 10 2017 11:19:10 GMT-0700 (PDT)
[debug] <<< HTTP RESPONSE 200
[debug] >>> HTTP REQUEST GET https://admin.firebase.com/v1/database/iotroniks/tokens  
 Wed May 10 2017 11:19:11 GMT-0700 (PDT)
[debug] <<< HTTP RESPONSE 200
[info] Starting Firebase development server...
[info] 
[info] Project Directory: /Users/ericlee/Documents/Projects/iotronik/firebase
[debug] >>> HTTP REQUEST GET https://cloudresourcemanager.googleapis.com/v1/projects/iotroniks  
 Wed May 10 2017 11:19:12 GMT-0700 (PDT)
[debug] <<< HTTP RESPONSE 200
[debug] >>> HTTP REQUEST GET https://mobilesdk-pa.googleapis.com/v1/projects/512273627061/clients/_:getWebAppConfig  
 Wed May 10 2017 11:19:13 GMT-0700 (PDT)
[debug] <<< HTTP RESPONSE 200
[info] Public Directory: public
[info] 
[info] Server listening at: http://localhost:5000
