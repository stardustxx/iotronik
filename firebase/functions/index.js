const functions = require('firebase-functions');
const admin = require('firebase-admin');

const topicPathNewIncident = 'incident-new';

// // Create and Deploy Your First Cloud Functions
// // https://firebase.google.com/docs/functions/write-firebase-functions
//
// exports.helloWorld = functions.https.onRequest((request, response) => {
//  response.send("Hello from Firebase!");
// });

admin.initializeApp(functions.config().firebase);

// Add new incident when new occureence is added
exports.addIncident = functions.database.ref('/occurrence/{occurrenceId}').onWrite((event) => {
  const originalValue = event.data.val();
  console.log('Adding new incident');
  console.log("originalValue", originalValue);
  const incidentRef = admin.database().ref('incident');
  const requiredTimeElapsed = 1000 * 60 * 5;
  
  if (originalValue) {
    // incidentRef.push().set(originalValue);
    incidentRef.orderByChild('time').limitToLast(1).once('value').then((snapshot) => {
      if (snapshot.val()) {
        // Incident reference list is not empty
        console.log("Snapshot exists: true");
        // Looping through the snapshot iterables, but there should only be 1 due to earlier filtering
        snapshot.forEach((childSnapShot) => {
          console.log("childSnapShot", childSnapShot.val());
          // Checking if the occurrence time and incident time is smaller than pre-set time interval
          // If so then add this occurrence to that incident
          if (originalValue.time - childSnapShot.val().time < requiredTimeElapsed) {
            console.log("Within time: true");
            incidentRef.child(childSnapShot.key).child('data').push().set({
              image: originalValue.image
            });
          } else {
            console.log("Within time: false");
            addNewIncident(incidentRef, originalValue);
          }
        });
      } else {
        // Incident reference list is empty, then just add new incident
        console.log("Snapshot exists: false");
        addNewIncident(incidentRef, originalValue);
      }
    }).catch((error) => {
      console.error("Incident Ref", error);
    });
  } else {
    console.log("Occurrence value does not exist");
  }
});

/**
 * Create new incident entry and add data
 * 
 * @param {any} reference     Firebase database reference of incident node
 * @param {any} originalValue Occurrence value
 */
function addNewIncident(reference, originalValue) {
  console.log("addNewIncident ran");
  const newKey = reference.push().key;
  const newIncidentRef = reference.child(newKey);
  newIncidentRef.set({
    time: originalValue.time
  }).then((success) => {
    newIncidentRef.child("data").push().set(originalValue);
    sendNotification();
  });
}

// Add new occurrence when a new image is created
exports.addOccurrence = functions.storage.object().onChange(event => {
  console.log('Adding new occurrence');
  const object = event.data;
  const filePath = object.name;
  const contentType = object.contentType;
  const resourceState = object.resourceState;
  const metageneration = object.metageneration;

  if (resourceState === 'exists' && metageneration == 1 && contentType.startsWith('image/')) {
    const occurrenceRef = admin.database().ref('occurrence');
    const dateNow = Date.now();
    occurrenceRef.push().set({
      time: dateNow,
      image: filePath
    });
  }

});

/**
 * Send a push notification
 * 
 */
function sendNotification() {
  const payload = {
    data: {
      "title": "Iotronik",
      "body": "You have a new image"
    },
    "notification": {
      "title": "Iotroink",
      "body": "You have a new image"
    }
  }
  admin.messaging().sendToTopic(topicPathNewIncident, payload).then((response) => {
    console.log("Successfully sent message:", response);
  }).catch((error) => {
    console.error("Error sending message:", error);
  });
}