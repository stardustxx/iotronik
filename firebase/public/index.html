<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Firebase Page</title>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
  </head>
  <body>
    <h1>Home Page</h1>
    <form id="test-form">
      <input id="btnFileSelect" type="file" name="" value="Select file">
      <button type="submit">Submit</button>
    </form>
    <script src="index.js"></script>
  </body>
</html>