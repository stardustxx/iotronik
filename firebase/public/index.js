const fileForm = document.querySelector("#test-form");
const btnFileSelect = document.querySelector("#btnFileSelect");

fileForm.onsubmit = (event) => {
  event.preventDefault();

  let file = btnFileSelect.files[0];
  
  if (file) {
    console.log("File", file);

    submitData(file);
  } else {
    console.log("Please select a file to upload");
  }

}

const submitData = (file) => {
  let formData = new FormData();
  formData.append('file', file);
  formData.append('name', `time_${Date.now()}`);

  let init = { 
    method: 'POST',
    cache: 'default',
    body: formData
  };

  fetch("http://localhost:3000/occurrence", init).then((response) => {
    return response.json();
  }).then((result) => {
    console.log(result);
  });
}