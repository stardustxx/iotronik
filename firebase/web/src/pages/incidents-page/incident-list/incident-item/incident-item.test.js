import React, { Component } from "react";
import IncidentItem from "./incident-item";
import renderer from "react-test-renderer";
import MuiThemeProvider from "material-ui/styles/MuiThemeProvider";
import Helpers from "../../../../helpers";
import "firebase/storage";

test("Incident item is rendered", () => {
  Helpers.initializeFirebase();
  const component = renderer.create(
    <MuiThemeProvider>
      <IncidentItem date="1497149501713" image="file.jpg_1497149499937" />
    </MuiThemeProvider>
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
