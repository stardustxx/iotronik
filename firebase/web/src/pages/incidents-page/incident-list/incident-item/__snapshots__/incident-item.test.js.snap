// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Incident item is rendered 1`] = `
<div>
  <div>
    <span
      disabled={undefined}
      href={undefined}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      onTouchTap={[Function]}
      open={null}
      style={
        Object {
          "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
          "background": "none",
          "backgroundColor": null,
          "border": 10,
          "boxSizing": "border-box",
          "color": "rgba(0, 0, 0, 0.87)",
          "cursor": "pointer",
          "display": "block",
          "fontFamily": "Roboto, sans-serif",
          "fontSize": 16,
          "fontWeight": "inherit",
          "lineHeight": "16px",
          "margin": 0,
          "muiPrepared": true,
          "outline": "none",
          "padding": 0,
          "position": "relative",
          "textDecoration": "none",
          "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
          "verticalAlign": null,
          "zIndex": 1,
        }
      }
      tabIndex={0}
    >
      <div
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
      >
        <div
          style={
            Object {
              "marginLeft": 0,
              "muiPrepared": true,
              "paddingBottom": 16,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 20,
              "position": "relative",
            }
          }
        >
          <div
            style={
              Object {
                "muiPrepared": true,
              }
            }
          >
            1497149501713
          </div>
          <img
            alt=""
            className="activity-item-image"
            src={null}
            style={
              Object {
                "WebkitBoxOrient": null,
                "WebkitLineClamp": null,
                "color": "rgba(0, 0, 0, 0.54)",
                "display": null,
                "fontSize": 14,
                "height": "100%",
                "lineHeight": "16px",
                "margin": 0,
                "marginTop": 4,
                "maxHeight": "450px",
                "maxWidth": "450px",
                "muiPrepared": true,
                "overflow": "hidden",
                "textOverflow": "ellipsis",
                "whiteSpace": "nowrap",
                "width": "100%",
              }
            }
          />
        </div>
      </div>
    </span>
  </div>
</div>
`;
